include/ibamr/LevelSetUtilities.h

```diff
@@ -0,0 +1,420 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2017 - 2020 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_LevelSetUtilities
+#define included_IBAMR_LevelSetUtilities
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibamr/config.h>
+
+#include "ibamr/LSInitStrategy.h"
+
+#include "tbox/DescribedClass.h"
+#include "tbox/Pointer.h"
+#include "tbox/Serializable.h"
+
+#include <limits>
+#include <string>
+
+namespace SAMRAI
+{
+namespace hier
+{
+template <int DIM>
+class PatchHierarchy;
+}
+namespace pdat
+{
+template <int DIM, class TYPE>
+class CellVariable;
+} // namespace pdat
+namespace tbox
+{
+class Database;
+} // namespace tbox
+} // namespace SAMRAI
+
+namespace IBAMR
+{
+class AdvDiffHierarchyIntegrator;
+} // namespace IBAMR
+
+namespace IBTK
+{
+class HierarchyMathOps;
+}
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class LevelSetUtilities provides implementation of some helper functions
+ * to enforce mass/volume conservation of phases.
+ */
+
+class LevelSetUtilities
+{
+public:
+    /*!
+     * \brief A lightweight class to hold the level set variable and the associated hierarchy
+     * integrator (AdvDiffHierarchyIntegrator).
+     */
+    class LevelSetContainer : public virtual SAMRAI::tbox::DescribedClass
+    {
+    public:
+        /*!
+         * \brief Constructor of the class.
+         */
+        LevelSetContainer(SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> adv_diff_integrator,
+                          SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_var)
+            : d_adv_diff_integrator(adv_diff_integrator)
+        {
+            d_ls_vars.emplace_back(ls_var);
+            return;
+        } // LevelSetContainer
+
+        LevelSetContainer(SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> adv_diff_integrator,
+                          std::vector<SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > > ls_vars)
+            : d_adv_diff_integrator(adv_diff_integrator), d_ls_vars(std::move(ls_vars))
+        {
+            // return;
+        } // LevelSetContainer
+
+        void setInterfaceHalfWidth(double ncells)
```

Comment:
Is there a reason not to include this in the constructor too? The other variables are essentially read only --- should this one be handled similarly?

Reply:
Yes. `ncells` is not a property of level set function, but rather the algorithm utilizing the level set variable. They can change them during the simulation.

Reply:
Also, is there source doe documentation somewhere for how this parameter is used?

Reply:
I will add a comment in .h file, where `ncells` gets used.